### PR TITLE
src: use image digest

### PR DESCRIPTION
**Pull request by zroubalik**
Signed-off-by: Zbynek Roubalik <zroubali@redhat.com>

Using image Digest in the Knative Service. Haven't moved to Docker Go SDK, because with this library it is [quite hard ](https://github.com/moby/moby/issues/39377#issuecomment-503589708) to use some authentication 💩 

We can try to migrate to this lib post Dev Preview. Issue created: https://github.com/boson-project/faas/issues/179

Fixes: #141 
Fixes: #140

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -414,7 +415,14 @@ func (c *Client) Deploy(path string) (err error) {
...
 	// Push the image for the named service to the configured registry
-	if err = c.pusher.Push(f); err != nil {
+	imageDigest, err := c.pusher.Push(f)
+	if err != nil {
+		return
+	}
+
+	// Store the produced image Digest in the config
+	f.ImageDigest = imageDigest
+	if err = writeConfig(f); err != nil {
```
Comment (zroubalik) [inline client.go:425]: It is not needed to store the digest curretnly (the push is immediately followed by deploy), so I can remove this part. But I thought we might want to use that in the future, if we introduce a ways how to fire `push` `deploy-only` etc (as was discussed on Slack).

File: docker/runner.go [hunk 2]
```diff
@@ -30,8 +30,8 @@ func (n *Runner) Run(f faas.Function) error {
 		return errors.New("please install 'docker'")
 	}
 
-	if f.Image == "" {
-		return errors.New("Function has no associated image.  Has it been built?")
+	if f.Image == "" || f.ImageWithDigest() == "" {
+		return errors.New("Function has no associated image and SHA356 hash. Has it been built?")
```
Comment (matejvasek) [inline docker/runner.go:34]: SHA356?

Comment (zroubalik) [reply docker/runner.go:34]: When 256 is not enough 😅  Good catch, thanks!

Comment (zroubalik) [reply docker/runner.go]: Fixed

File: go.mod [hunk 3]
```diff
@@ -4,6 +4,7 @@ go 1.14
 
 require (
 	github.com/buildpacks/pack v0.14.0
+	github.com/docker/docker v1.13.1
```
Comment (dsimansk) [inline go.mod:7]: I don't see it used anywhere, just be aware that the version will be overridden by `replace` block below. In addition I was under impression that `v1.13.1` doesn't work with current buildpack version. 

Dependencies, dependencies, ... :)

Comment (zroubalik) [reply go.mod]: Good catch, thanks! This is a leftover from previous iteration.  Fixed

**Review by lance — APPROVED** [review]

**Pull request merged**